proxy/coco/src/peer/subroutines.rs

```diff
@@ -0,0 +1,299 @@
+    fn spawn_command(&self, cmd: Command) -> SpawnAbortable<()> {
+        match cmd {
+            Command::Announce => SpawnAbortable::new(announce(
+                self.state.clone(),
+                self.store.clone(),
+                self.input_sender.clone(),
+            )),
+            Command::Control(control_command) => match control_command {
+                ControlCommand::Respond(respond_command) => {
+                    SpawnAbortable::new(control_respond(respond_command))
+                },
+            },
+            Command::SyncPeer(peer_id) => {
+                SpawnAbortable::new(sync(self.state.clone(), peer_id, self.input_sender.clone()))
+            },
+            Command::StartSyncTimeout(sync_period) => {
+                SpawnAbortable::new(start_sync_timeout(sync_period, self.input_sender.clone()))
+            },
+            Command::Request(RequestCommand::Query(urn)) => {
+                SpawnAbortable::new(query(urn, self.state.clone(), self.input_sender.clone()))
+            },
+            Command::Request(RequestCommand::Clone(url)) => {
+                SpawnAbortable::new(clone(url, self.state.clone(), self.input_sender.clone()))
+            },
+        }
+    }
+}
+
+impl Drop for Subroutines {
+    fn drop(&mut self) {
+        for task in self.pending_tasks.iter_mut() {
+            task.abort()
+        }
+    }
+}
+
+impl Future for Subroutines {
+    type Output = Result<(), spawn_abortable::Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        // Drain the task queue.
+        loop {
+            match self.pending_tasks.poll_next_unpin(cx) {
+                Poll::Ready(Some(Err(e))) => {
+                    log::warn!("error in spawned subroutine task: {:?}", e);
+                    return Poll::Ready(Err(e));
+                },
+                Poll::Ready(Some(Ok(()))) => continue,
+                // Either pending, or FuturesUnordered thinks it's done, but
+                // we'll enqueue new tasks below
+                Poll::Ready(None) | Poll::Pending => break,
+            }
+        }
+
+        // Drain all pending inputs, feed them to the [`RunState`] and execute the returned
+        // commands as async tasks.
+        loop {
+            match self.inputs.poll_next_unpin(cx) {
+                Poll::Ready(Some(input)) => {
+                    log::debug!("handling subroutine input: {:?}", input);
+
+                    if let Some(event) = Event::maybe_from(&input) {
+                        // Ignore if there are no subscribers.
+                        self.subscriber.send(event).ok();
+                    }
+
+                    for cmd in self.run_state.transition(input) {
+                        let task = self.spawn_command(cmd);
+
+                        self.pending_tasks.push(task);
+                    }
+                },
+                Poll::Ready(None) => return Poll::Ready(Ok(())),
```

Comment:
Is it correct that we want to abort all running tasks once the input stream stops?

Reply:
I think so, because the only way the stream can signal `Ready` is because all the channels the `SelectAll` is closing over return `Ready`. This can only happen when all their senders have been dropped.

Reply:
I’m just wondering whether we want to wait for all the tasks to be finished.